File: LeetCode/205.Isomorphic_Strings/205.Isomorphic_Strings.swift
class Solution {
    func isIsomorphic(_ s: String, _ t: String) -> Bool {
        // edge case
        if s.count != t.count {
            return false
        }

        var sMapping: [Character: Character] = [:]
        var tMapping: [Character: Character] = [:]

        for (ss, tt) in zip(Array(s), Array(t)) {
            // check s -> t
            if let a = sMapping[ss], a != tt {
                return false
            }
            // check t -> s
            if let b = tMapping[tt], b != ss {
                return false
            }

            // build s->t && t -> s
            if sMapping[ss] == nil && tMapping[tt] == nil {
                sMapping[ss] = tt
                tMapping[tt] = ss
            }
        }

        return true
    }
}

class Solution {
    func isIsomorphic(_ s: String, _ t: String) -> Bool {
        // edge case
        if s.count != t.count {
            return false
        }

        var mapST = Array(repeating: -1, count: 256)
        var mapTS = Array(repeating: -1, count: 256)

        for (ss, tt) in zip(s.utf8, t.utf8) {
            let si = Int(ss), ti = Int(tt)
            // build mapping
            if mapST[si] == -1 && mapTS[ti] == -1 {
                mapST[si] = ti
                mapTS[ti] = si
            } else if mapST[si] != ti || mapTS[ti] != si {
                return false
            }
        }

        return true
    }
}

File: LeetCode/205.Isomorphic_Strings/205.Isomorphic_Strings.py
#!/usr/bin/env python3


class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        # build map char in s can map to char in s
        s_to_t, t_to_s = {}, {}
        for a, b in zip(s, t):
            if a not in s_to_t and b not in t_to_s:
                s_to_t[a] = b
                t_to_s[b] = a
            elif a in s_to_t and b in t_to_s:
                if s_to_t[a] == b and t_to_s[b] == a:
                    continue
                else:
                    return False
            else:
                return False

        return True
